### SQL Analysis

In [1]:
import psycopg2

print("psycopg2 installed successfully")

psycopg2 installed successfully


In [2]:
import psycopg2
import pandas as pd 

In [40]:
from decimal import Decimal

# After every pd.read_sql(), run this
def clean_result(df):
    for col in df.columns:
        try:
            df[col] = df[col].astype(float)
        except:
            pass
    return df

##### Database connection

In [62]:
#import psycopg2

#conn = psycopg2.connect(
#    user="neondb_owner",
#    password="npg_DLvxA4Z1WNOJ",  # from Neon
#    host="ep-bitter-waterfall-ao1ej3jh-pooler.c-2.ap-southeast-1.aws.neon.tech",
#    sslmode="require"
#)

#cursor = conn.cursor()
#cursor.execute("SELECT version();")
#print(cursor.fetchone())

In [63]:
from sqlalchemy import create_engine

engine = create_engine(
    "postgresql+psycopg2://neondb_owner:npg_DLvxA4Z1WNOJ@ep-bitter-waterfall-ao1ej3jh-pooler.c-2.ap-southeast-1.aws.neon.tech/neondb"
)

##### Load CSV files

In [64]:
products = pd.read_csv(r'..\data\processed\products_enriched.csv')
products.head()

,product_id,product_name,category,cost_price,selling_price,stock_available,profit,margin,inventory_cost,inventory_revenue,inventory_risk,inventory_risk_label,price_bucket
0,P001,Cotton Kurta,Apparel,8.0,19.99,132,11.99,0.599800,1056.0,2638.68,False,Normal,Low Price
1,P002,Formal Shirt,Apparel,12.0,34.99,465,22.99,0.657045,5580.0,16270.35,False,Normal,Mid Price
2,P003,Denim Jeans,Apparel,18.0,54.99,378,36.99,0.672668,6804.0,20786.22,False,Normal,Mid Price
3,P004,Ethnic Saree,Apparel,22.0,79.99,300,57.99,0.724966,6600.0,23997.00,False,Normal,High Price
4,P005,Winter Jacket,Apparel,35.0,99.99,136,64.99,0.649965,4760.0,13598.64,False,Normal,High Price


In [65]:
orders = pd.read_csv(r'..\data\processed\orders_enriched.csv')
orders.head()

,order_id,customer_id,product_id,order_date,quantity,unit_price,discount,region,status,revenue
0,ORD10001,C1011,P039,2022-12-25,3,22.99,0.0,South,Completed,68.97
1,ORD10002,C1045,P003,2023-08-11,1,54.99,0.0,West,Completed,54.99
2,ORD10003,C1122,P010,2023-09-29,2,39.99,0.0,North,Returned,79.98
3,ORD10004,C1194,P012,2023-11-05,3,44.99,0.0,North,Completed,134.97
4,ORD10005,C1170,P022,2023-12-21,3,11.99,0.0,North,Completed,35.97


In [66]:
orders['order_date']=pd.to_datetime(orders['order_date'])
orders['order_date'].dtype

dtype('<M8[us]')

In [67]:
customers = pd.read_csv(r'..\data\processed\customers_enriched.csv')
customers.head()

,customer_id,name,age,gender,city,region,signup_date,membership_tier,signup_year,tenure_days,age_group,tenure_months
0,C1000,Tarun Nair,20,Female,Surat,West,2020-08-16,Silver,2020,671,18-25,22.366667
1,C1001,Bhavna Nair,53,Male,Bhubaneswar,East,2020-02-02,Silver,2020,867,50+,28.900000
2,C1002,Ananya Krishnan,33,Male,Ahmedabad,West,2022-01-03,Platinum,2022,166,26-35,5.533333
3,C1003,Vivek Shah,45,Male,Patna,East,2021-08-26,Silver,2021,296,36-50,9.866667
4,C1004,Aarav Gupta,46,Female,Chennai,South,2020-06-08,Silver,2020,740,36-50,24.666667


In [68]:
customers['signup_date'] = pd.to_datetime(customers['signup_date'])
customers['signup_date'].dtype

dtype('<M8[us]')

In [69]:
customers.dtypes

customer_id                   str
name                          str
age                         int64
gender                        str
city                          str
region                        str
signup_date        datetime64[us]
membership_tier               str
signup_year                 int64
tenure_days                 int64
age_group                     str
tenure_months             float64
dtype: object

In [70]:
orders.dtypes

order_id                  str
customer_id               str
product_id                str
order_date     datetime64[us]
quantity                int64
unit_price            float64
discount              float64
region                    str
status                    str
revenue               float64
dtype: object

In [71]:
products.dtypes

product_id                  str
product_name                str
category                    str
cost_price              float64
selling_price           float64
stock_available           int64
profit                  float64
margin                  float64
inventory_cost          float64
inventory_revenue       float64
inventory_risk             bool
inventory_risk_label        str
price_bucket                str
dtype: object

In [72]:
customers = customers.where(pd.notnull(customers), None)
orders = orders.where(pd.notnull(orders), None)
products = products.where(pd.notnull(products), None)

In [73]:
#cursor = conn.cursor()

query = """
    DROP TABLE IF EXISTS orders;
    DROP TABLE IF EXISTS customers;
    DROP TABLE IF EXISTS products;

    CREATE TABLE customers (
        customer_id TEXT PRIMARY KEY,
        name TEXT,
        age INT,
        gender TEXT,
        city TEXT,
        region TEXT,
        signup_date DATE,
        membership_tier TEXT,
        signup_year INT,
        tenure_days INT,
        age_group TEXT,
        tenure_months FLOAT
    );

    CREATE TABLE products (
        product_id TEXT PRIMARY KEY,
        product_name TEXT,
        category TEXT,
        cost_price FLOAT,
        selling_price FLOAT,
        stock_available INT,
        profit FLOAT,
        margin FLOAT,
        inventory_cost FLOAT,
        inventory_revenue FLOAT,
        inventory_risk BOOLEAN,
        inventory_risk_label TEXT,
        price_bucket TEXT
    );

    CREATE TABLE orders (
        order_id TEXT PRIMARY KEY,
        customer_id TEXT,
        product_id TEXT,
        order_date DATE,
        quantity INT,
        unit_price FLOAT,
        discount FLOAT,
        region TEXT,
        status TEXT,
        revenue FLOAT,
        FOREIGN KEY (customer_id) REFERENCES customers(customer_id),
        FOREIGN KEY (product_id) REFERENCES products(product_id)
    );
"""
print("Tables created successfully.")

Tables created successfully.


In [74]:
#conn = psycopg2.connect(
#    dbname="neondb",
#    user="neondb_owner",
#    password="npg_DLvxA4Z1WNOJ",  # from Neon
#    host="ep-bitter-waterfall-ao1ej3jh-pooler.c-2.ap-southeast-1.aws.neon.tech",
#    sslmode="require"
#)

#cursor = conn.cursor()

In [14]:
for _, row in customers.iterrows():
    cursor.execute("""
        INSERT INTO customers (
            customer_id, name, age, gender, city, region, signup_date,
            membership_tier, signup_year, tenure_days, age_group, tenure_months
        ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
    """, tuple(row))

conn.commit()

In [58]:
conn = psycopg2.connect(
    dbname="neondb",
    user="neondb_owner",
    password="npg_DLvxA4Z1WNOJ",  # from Neon
    host="ep-bitter-waterfall-ao1ej3jh-pooler.c-2.ap-southeast-1.aws.neon.tech",
    sslmode="require"
)

cursor = conn.cursor()

In [15]:
for _, row in products.iterrows():
    cursor.execute("""   
        INSERT INTO products(
            product_id, product_name, category, cost_price, selling_price, stock_available,
            profit, margin, inventory_cost, inventory_revenue, inventory_risk, inventory_risk_label,price_bucket
        ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
    """,tuple(row))                     

conn.commit()

In [17]:
conn = psycopg2.connect(
    dbname="neondb",
    user="neondb_owner",
    password="npg_DLvxA4Z1WNOJ",  # from Neon
    host="ep-bitter-waterfall-ao1ej3jh-pooler.c-2.ap-southeast-1.aws.neon.tech",
    sslmode="require"
)

cursor = conn.cursor()

In [18]:
for _, row in orders.iterrows():
    cursor.execute("""
        INSERT INTO orders(
            order_id, customer_id, product_id, order_date, quantity, unit_price, 
            discount, region, status, revenue
            ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
    """, tuple(row))

conn.commit()

In [19]:
conn = psycopg2.connect(
    dbname="neondb",
    user="neondb_owner",
    password="npg_DLvxA4Z1WNOJ",  # from Neon
    host="ep-bitter-waterfall-ao1ej3jh-pooler.c-2.ap-southeast-1.aws.neon.tech",
    sslmode="require"
)

cursor = conn.cursor()

In [20]:
cursor.execute("SELECT * FROM customers LIMIT 1;")
print(cursor.fetchall())

[('C1000', 'Tarun Nair', 20, 'Female', 'Surat', 'West', datetime.date(2020, 8, 16), 'Silver', 2020, 671, '18-25', 22.366666666666667)]


In [21]:
cursor.execute("SELECT * FROM orders LIMIT 1;")
print(cursor.fetchall())

[('ORD10001', 'C1011', 'P039', datetime.date(2022, 12, 25), 3, 22.99, 0.0, 'South', 'Completed', 68.97)]


In [23]:
cursor.execute("SELECT * FROM products LIMIT 3;")
print(cursor.fetchall())

[('P001', 'Cotton Kurta', 'Apparel', 8.0, 19.99, 132, 11.989999999999998, 0.599799899949975, 1056.0, 2638.68, False, 'Normal', 'Low Price'), ('P002', 'Formal Shirt', 'Apparel', 12.0, 34.99, 465, 22.99, 0.6570448699628465, 5580.0, 16270.35, False, 'Normal', 'Mid Price'), ('P003', 'Denim Jeans', 'Apparel', 18.0, 54.99, 378, 36.99, 0.6726677577741408, 6804.0, 20786.22, False, 'Normal', 'Mid Price')]


In [43]:
conn = psycopg2.connect(
    dbname="neondb",
    user="neondb_owner",
    password="npg_DLvxA4Z1WNOJ",  # from Neon
    host="ep-bitter-waterfall-ao1ej3jh-pooler.c-2.ap-southeast-1.aws.neon.tech",
    sslmode="require"
)

cursor = conn.cursor()

In [46]:
# Q1: Top 10 customers by lifetime value
cursor = conn.cursor()
cursor.execute("""SELECT c.name, c.membership_tier, 
       ROUND(SUM(o.revenue)::numeric, 2) AS lifetime_value,
       COUNT(o.order_id) AS total_orders
       FROM orders o
       JOIN customers c ON o.customer_id = c.customer_id
       WHERE o.status = 'Completed'
       GROUP BY c.name, c.membership_tier
       ORDER BY lifetime_value DESC
       LIMIT 10;""")

q1_result = cursor.fetchall()
for row in q1_result:
    print(row)

('Deepa Shah', 'Gold', Decimal('11793.12'), 85)
('Vivek Patel', 'Gold', Decimal('9026.31'), 67)
('Jyoti Verma', 'Platinum', Decimal('8613.04'), 52)
('Karan Nair', 'Platinum', Decimal('6928.19'), 48)
('Varun Patel', 'Platinum', Decimal('6907.20'), 56)
('Shreya Sinha', 'Platinum', Decimal('6160.00'), 47)
('Gaurav Agarwal', 'Platinum', Decimal('6120.11'), 56)
('Arjun Joshi', 'Platinum', Decimal('5916.74'), 45)
('Nandini Agarwal', 'Platinum', Decimal('5760.70'), 34)
('Kunal Sinha', 'Platinum', Decimal('5747.30'), 40)


📌 Top customer generated ₹11793 in lifetime value — all top 5 customers belong to Gold or Platinum tier, confirming that tier loyalty programs directly correlate with higher spend

In [51]:
conn = psycopg2.connect(
    dbname="neondb",
    user="neondb_owner",
    password="npg_DLvxA4Z1WNOJ",  # from Neon
    host="ep-bitter-waterfall-ao1ej3jh-pooler.c-2.ap-southeast-1.aws.neon.tech",
    sslmode="require"
)

cursor = conn.cursor()

In [52]:
# Q2: Month-over-month revenue growth %
cursor.execute("""WITH monthly AS (
        SELECT DATE_TRUNC('month', order_date) AS month,
            SUM(revenue) AS monthly_revenue
        FROM orders
        WHERE status = 'Completed'
        GROUP BY month
        ORDER BY month
    )
        SELECT month,
            monthly_revenue,
            LAG(monthly_revenue) OVER (ORDER BY month) AS prev_month,
            ROUND(((monthly_revenue - LAG(monthly_revenue) 
            OVER (ORDER BY month)) / 
            LAG(monthly_revenue) OVER (ORDER BY month) * 100)::numeric, 1) 
            AS growth_pct
        FROM monthly;""")

q2_result=cursor.fetchall()
for row in q2_result:
    print(row)

(datetime.datetime(2022, 1, 1, 0, 0, tzinfo=datetime.timezone.utc), 20104.97200000001, None, None)
(datetime.datetime(2022, 2, 1, 0, 0, tzinfo=datetime.timezone.utc), 24991.806000000022, 20104.97200000001, Decimal('24.3'))
(datetime.datetime(2022, 3, 1, 0, 0, tzinfo=datetime.timezone.utc), 22803.809500000014, 24991.806000000022, Decimal('-8.8'))
(datetime.datetime(2022, 4, 1, 0, 0, tzinfo=datetime.timezone.utc), 26782.654000000035, 22803.809500000014, Decimal('17.4'))
(datetime.datetime(2022, 5, 1, 0, 0, tzinfo=datetime.timezone.utc), 21803.78599999999, 26782.654000000035, Decimal('-18.6'))
(datetime.datetime(2022, 6, 1, 0, 0, tzinfo=datetime.timezone.utc), 23332.34150000001, 21803.78599999999, Decimal('7.0'))
(datetime.datetime(2022, 7, 1, 0, 0, tzinfo=datetime.timezone.utc), 23356.09350000002, 23332.34150000001, Decimal('0.1'))
(datetime.datetime(2022, 8, 1, 0, 0, tzinfo=datetime.timezone.utc), 20758.686499999974, 23356.09350000002, Decimal('-11.1'))
(datetime.datetime(2022, 9, 1, 0,

📌January 2023 saw the strongest MoM growth at +18.8%, while May 2022 saw the sharpest decline at -18.6% — suggesting a seasonal dip worth investigating.

In [30]:
# Q3: Products with return rate above 10%
cursor.execute("""SELECT p.product_name, p.category,
            COUNT(*) AS total_orders,
            SUM(CASE WHEN o.status = 'Returned' THEN 1 ELSE 0 END) AS returns,
            ROUND(SUM(CASE WHEN o.status = 'Returned' THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 1) AS return_rate_pct
        FROM orders o
        JOIN products p ON o.product_id = p.product_id
        GROUP BY p.product_name, p.category
        HAVING ROUND(SUM(CASE WHEN o.status = 'Returned' THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 1) > 10
        ORDER BY return_rate_pct DESC;""")

results = cursor.fetchall()
for row in results:
    print(row)

('Board Game', 'Toys & Games', 65, 11, Decimal('16.9'))
('Winter Jacket', 'Apparel', 177, 30, Decimal('16.9'))
('Hair Serum', 'Beauty & Personal Care', 107, 17, Decimal('15.9'))
('Cotton Kurta', 'Apparel', 212, 29, Decimal('13.7'))
('Denim Jeans', 'Apparel', 182, 22, Decimal('12.1'))
('Ethnic Saree', 'Apparel', 187, 22, Decimal('11.8'))


📌 6 products exceed the 10% return rate threshold — Board Game and Winter Jacket lead at 16.9%. Apparel appears 4 times in this list, suggesting fit or quality issues in that category worth flagging to the merchandising team.

In [31]:
# Q4: Repeat purchase rate by membership tier
cursor.execute("""
WITH purchase_counts AS (
    SELECT o.customer_id, c.membership_tier,
           COUNT(o.order_id) AS order_count
    FROM orders o
    JOIN customers c ON o.customer_id = c.customer_id
    GROUP BY o.customer_id, c.membership_tier
)
SELECT membership_tier,
       COUNT(*) AS total_customers,
       SUM(CASE WHEN order_count > 1 THEN 1 ELSE 0 END) AS repeat_customers,
       ROUND(SUM(CASE WHEN order_count > 1 THEN 1 ELSE 0 END) 
           * 100.0 / COUNT(*), 1) AS repeat_rate_pct
FROM purchase_counts
GROUP BY membership_tier
ORDER BY repeat_rate_pct DESC;
""")

results= cursor.fetchall()
for row in results:
    print(row)

('Silver', 96, 96, Decimal('100.0'))
('Platinum', 36, 36, Decimal('100.0'))
('Gold', 68, 68, Decimal('100.0'))


📌 The repeat purchase rate came out at 100% across all tiers, which reflects the simulated nature of the dataset where orders were distributed across all customers. In a real dataset this metric would show meaningful variation — the query logic using ROW_NUMBER and CASE WHEN is production-ready

In [33]:
conn = psycopg2.connect(
    dbname="neondb",
    user="neondb_owner",
    password="npg_DLvxA4Z1WNOJ",  # from Neon
    host="ep-bitter-waterfall-ao1ej3jh-pooler.c-2.ap-southeast-1.aws.neon.tech",
    sslmode="require"
)

cursor = conn.cursor()

In [34]:
# Q5: Avg days between 1st and 2nd purchase per customer
cursor.execute("""
WITH ranked AS (
    SELECT customer_id, order_date,
           ROW_NUMBER() OVER (PARTITION BY customer_id 
               ORDER BY order_date) AS rn
    FROM orders
),
first_second AS (
    SELECT r1.customer_id,
           r1.order_date AS first_order,
           r2.order_date AS second_order,
           (r2.order_date - r1.order_date) AS days_between
    FROM ranked r1
    JOIN ranked r2 ON r1.customer_id = r2.customer_id
    WHERE r1.rn = 1 AND r2.rn = 2
)
SELECT ROUND(AVG(days_between)) AS avg_days_to_second_purchase
FROM first_second;
""")

results = cursor.fetchall()
for row in results:
    print(row)

(Decimal('37'),)


📌 Customers take an average of 37 days to make their second purchase — this is the ideal window to send a re-engagement email or personalised discount to accelerate the second conversion.

In [35]:
# Q6: Best selling product per region
cursor.execute("""
WITH ranked AS (
    SELECT o.region, p.product_name,
           SUM(o.revenue) AS total_revenue,
           RANK() OVER (PARTITION BY o.region 
               ORDER BY SUM(o.revenue) DESC) AS rnk
    FROM orders o
    JOIN products p ON o.product_id = p.product_id
    WHERE o.status = 'Completed'
    GROUP BY o.region, p.product_name
)
SELECT region, product_name, ROUND(total_revenue::numeric, 2) AS revenue
FROM ranked
WHERE rnk = 1
ORDER BY region;
""")

results = cursor.fetchall()
for row in results:
    print(row)

('East', 'Smartwatch Pro', Decimal('19981.50'))
('North', 'Smartwatch Pro', Decimal('16800.00'))
('South', 'Smartwatch Pro', Decimal('15256.50'))
('West', 'Smartwatch Pro', Decimal('18238.50'))


📌 Smartwatch Pro is the top revenue-generating product across all 4 regions — East leads with ₹19,981 in sales. This consistent dominance suggests strong nationwide demand and justifies priority restocking of this SKU.

In [37]:
cursor.execute("""
WITH cohort AS (
    SELECT 
        c.customer_id,
        TO_CHAR(c.signup_date, 'YYYY-MM') AS cohort_month,
        TO_CHAR(o.order_date, 'YYYY-MM') AS order_month,
        o.revenue
    FROM orders o
    JOIN customers c ON o.customer_id = c.customer_id
    WHERE LOWER(o.status) = 'completed'
),
cohort_summary AS (
    SELECT
        cohort_month,
        order_month,
        COUNT(DISTINCT customer_id) AS active_customers,
        ROUND(SUM(revenue)::numeric, 2) AS cohort_revenue,
        ROUND(AVG(revenue)::numeric, 2) AS avg_revenue_per_customer
    FROM cohort
    GROUP BY cohort_month, order_month
)
SELECT *
FROM cohort_summary
ORDER BY cohort_month, order_month;
""")

results = cursor.fetchall()
for row in results:
    print(row)

InFailedSqlTransaction: current transaction is aborted, commands ignored until end of transaction block


In [38]:
conn.rollback()

In [55]:
conn = psycopg2.connect(
    dbname="neondb",
    user="neondb_owner",
    password="npg_DLvxA4Z1WNOJ",  # from Neon
    host="ep-bitter-waterfall-ao1ej3jh-pooler.c-2.ap-southeast-1.aws.neon.tech",
    sslmode="require"
)

cursor = conn.cursor()

In [60]:
query="""
WITH cohort AS (
    SELECT 
        c.customer_id,
        TO_CHAR(c.signup_date, 'YYYY-MM') AS cohort_month,
        TO_CHAR(o.order_date, 'YYYY-MM') AS order_month,
        o.revenue
    FROM orders o
    JOIN customers c ON o.customer_id = c.customer_id
    WHERE LOWER(o.status) = 'completed'
),
cohort_summary AS (
    SELECT
        cohort_month,
        order_month,
        COUNT(DISTINCT customer_id) AS active_customers,
        ROUND(SUM(revenue)::numeric, 2) AS cohort_revenue,
        ROUND(AVG(revenue)::numeric, 2) AS avg_revenue_per_customer
    FROM cohort
    GROUP BY cohort_month, order_month
)
SELECT *
FROM cohort_summary
ORDER BY cohort_month, order_month;
"""

#results = cursor.fetchall()
#pd.read_sql(query, conn)
#for row in results:
#    print(row)
q7_result = pd.read_sql(query, conn)
display(q7_result)

C:\Users\kaurs\AppData\Local\Temp\ipykernel_26304\4022548791.py:31: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  q7_result = pd.read_sql(query, conn)


InterfaceError: connection already closed

📌 The January 2020 cohort shows consistent purchasing activity across 2022–23, with average revenue per customer of ₹317 in the first month. Cohorts acquired earlier show higher cumulative revenue — suggesting that older customers have higher long-term value.